### DISCUSSION TITLE
Frecency for args

### DISCUSSION THREAD

**Discussion by hakubo** [johnlindquist/kit#281] [Ideas]
Hey,
I've been toying with adding frecency to `arg`. For this quick demo I used frecency package from npm that is mainly targeting browsers, so this could be streamlined for kits purposes.

Basic idea is to bubble up selections that are used more frequently and/or recently.

I think such functionality could be built into kits `arg` that's why I'm sharing it. If you think this could be useful I can help with adding it.

Demo:

https://user-images.githubusercontent.com/1018759/118710911-3786a500-b81f-11eb-8379-ea7fab1e8082.mp4

```
// This does not work frecency uses export default - for demo purposes I copied and modified the script
const { default: Frecency } = await import("frecency");

const frecencyDb = db("frecency", {});

async function frecencyArg(name, choices) {
  let lastInput;

  const frecency = new Frecency({
    key: name,
    storageProvider: {
      getItem(key) {
        return frecencyDb.get(key).value();
      },
      setItem(key, value) {
        return frecencyDb.set(key, value).write();
      },
    },
  });

  const result = await arg(name, (query) => {
    lastInput = query;
    console.log(
      choices
        .filter((arg) => arg.includes(query))
        .map((arg) => ({ _id: arg, value: arg }))
    );

    return frecency
      .sort({
        searchQuery: query,
        results: choices
          .filter((arg) => arg.includes(query))
          .map((arg) => ({ _id: arg, value: arg })),
      })
      .map((arg) => arg.value);
  });

  frecency.save({
    searchQuery: lastInput,
    selectedId: result,
  });

  return result;
}

await frecencyArg("myArg", ["one", "two", "three"]);
```

**Comment by johnlindquist** [comment]
Very cool, definitely something I've been thinking about.

I think I'd like the API to be a traditional `sort`:

```js
//"recent" would be a "Frecency" sort fn
await arg({placeholder: "Select a name", sort: recent})

//vanilla sort
await arg({placeholder: "Select a name", sort: (a, b)=> a.name > b.name ? 1 : a.name === b.name ? 0 : -1})
```

This would require some internals work. I'll think about it for a while and maybe toy with a couple ideas.

**Reply by hakubo** [reply]
just 'sort' might not be enough, as you have to store what was selected with what search query. But I agree that this should go into options of 'arg'
